bindings/matrix-sdk-crypto-ffi/src/olm.udl

```diff
@@ -146,54 +146,87 @@ dictionary StartSasResult {
 };
 
-dictionary Sas {
-    string other_user_id;
-    string other_device_id;
-    string flow_id;
-    string? room_id;
-    boolean we_started;
-    boolean has_been_accepted;
-    boolean can_be_presented;
-    boolean supports_emoji;
-    boolean have_we_confirmed;
-    boolean is_done;
-    boolean is_cancelled;
-    CancelInfo? cancel_info;
+interface Sas {
+    string other_user_id();
+    string other_device_id();
+    string flow_id();
+    string? room_id();
+    boolean we_started();
+    boolean is_done();
+
+    OutgoingVerificationRequest? accept();
+    [Throws=CryptoStoreError]
+    ConfirmVerificationResult? confirm();
+    OutgoingVerificationRequest? cancel([ByRef] string cancel_code);
+
+    sequence<i32>? get_emoji_indices();
+    sequence<i32>? get_decimals();
+
+    void set_changes_listener(SasListener callback);
+};
+
+[Enum]
+interface SasState {
+  Started();
+  Accepted();
+  KeysExchanged(sequence<i32>? emojis, sequence<i32> decimals);
+  Confirmed();
+  Done();
+  Cancelled(CancelInfo cancel_info);
+};
+
+callback interface SasListener {
+    void on_change(SasState state);
 };
 
 dictionary ScanResult {
     QrCode qr;
     OutgoingVerificationRequest request;
 };
 
-dictionary QrCode {
-    string other_user_id;
-    string other_device_id;
-    string flow_id;
-    string? room_id;
-    boolean we_started;
-    boolean other_side_scanned;
-    boolean has_been_confirmed;
-    boolean reciprocated;
-    boolean is_done;
-    boolean is_cancelled;
-    CancelInfo? cancel_info;
-};
-
-dictionary VerificationRequest {
-    string other_user_id;
-    string? other_device_id;
-    string flow_id;
-    string? room_id;
-    boolean we_started;
-    boolean is_ready;
-    boolean is_passive;
-    boolean is_done;
-    boolean is_cancelled;
-    CancelInfo? cancel_info;
-    sequence<string>? their_methods;
-    sequence<string>? our_methods;
+interface QrCode {
+    string other_user_id();
+    string other_device_id();
+    string flow_id();
+    string? room_id();
+    boolean we_started();
+    boolean is_done();
+    boolean is_cancelled();
+    CancelInfo? cancel_info();
+
+    boolean reciprocated();
+    boolean has_been_scanned();
+
+    ConfirmVerificationResult? confirm();
+    OutgoingVerificationRequest? cancel([ByRef] string cancel_code);
+    string? generate_qr_code();
+};
+
+interface VerificationRequest {
+    string other_user_id();
+    string? other_device_id();
+    string flow_id();
+    string? room_id();
+    boolean we_started();
+    boolean is_ready();
+    boolean is_done();
+    boolean is_passive();
+    boolean is_cancelled();
+    CancelInfo? cancel_info();
+
+    sequence<string>? their_supported_methods();
+    sequence<string>? our_supported_methods();
+
+    OutgoingVerificationRequest? accept(sequence<string> methods);
+
+    [Throws=CryptoStoreError]
+    StartSasResult? start_sas_verification();
+
+    [Throws=CryptoStoreError]
+    QrCode? start_qr_verification();
```

Comment:
I find it confusing that it's called start_qr_verification. Makes me thinking that it's actually sending a m.verification.start. Maybe it could be create_qr_verification

Reply:
This reflects how the SAS API looks like. Again, I think it's expected that people don't know much about the `m.key.verification.*` events.

Though it is a bit confusing that the QR verification starts before a start event is sent out. I believe that this is addressed in v2 of the QR code verification.